README.md
# alfred-adcode

[![build](https://github.com/daolou/alfred-adcode/workflows/Create%20Alfred%20Workflow/badge.svg)](https://github.com/daolou/alfred-adcode/actions?query=workflow%3ACreate%20Alfred%20Workflow)

> 快速查询中国境内，行政区域（依赖 [高德开放平台](https://lbs.amap.com/api/webservice/guide/api/district)）

![alfred-adcode](./docs/assets/alfred-adcode.gif)

可以得到：

- 行政区名称
- 区域编码
- 城市编码
- 区域中心点
- 行政区级别
- 行政区边界坐标点

## 安装

_Requires [Node.js](https://nodejs.org) 14+ and the Alfred [Powerpack](https://www.alfredapp.com/powerpack/)._

1. [下载最新 workflow](https://github.com/daolou/alfred-adcode/releases/latest)
2. 双击 .alfredworkflow 文件即可进行安装

## 配置

打开配置，如图：![alfred-adcode-env](./docs/assets/alfred-adcode-env.png) 在环境变量 ADCODE_KEY，填入你自己申请的 key（需要预先申请，请看[申请 key](https://lbs.amap.com/dev/key)）

## 使用

1. 使用 `ad` 关键字 触发 workflow
2. 随后输入要查询的区域，可以是地区名（中文），也可以是 6 位数字的区域 code 码,比如 北京 或者 110000
3. 按下 <kbd>Enter</kbd> 或 <kbd>⌘ + c</kbd> 即可复制到粘贴板

## License

MIT © [KhaZix](https://github.com/daolou/alfred-adcode)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.3.1](https://github.com/daolou/alfred-adcode/compare/v1.3.0...v1.3.1) (2022-05-17)


### Bug Fixes

* 🐛 txiki not support import node_modules resolve ([76b6c15](https://github.com/daolou/alfred-adcode/commit/76b6c1590b1a7ec8e3c3e4cdf5d9f9084656307f))

## 1.3.0 (2022-05-17)


### Features

* ✨ exclude some file ([faf8ba3](https://github.com/daolou/alfred-adcode/commit/faf8ba31a8ee780922571bc234a93f461c6d96ea))
* ✨ optimize Drastically reduced size ([e4ef55b](https://github.com/daolou/alfred-adcode/commit/e4ef55b8f39442105380380fc21ba74592bc01bb))
* ✨ optimize Drastically reduced size ([3500e14](https://github.com/daolou/alfred-adcode/commit/3500e14c2f047b0bcbea100e466016c80b692184))


### Bug Fixes

* 🐛 ERR_MODULE_NOT_FOUND: node esm require extension .js and use cjs ([ca0d255](https://github.com/daolou/alfred-adcode/commit/ca0d255252051c20c218ae713c674bfa5c5fc93a))
* 🐛 ERR_MODULE_NOT_FOUND: node esm require extension .js and use cjs ([0010536](https://github.com/daolou/alfred-adcode/commit/00105369184158f7d3157563710600f1b6562530))
* 🐛 No such file or directory ([ea791f2](https://github.com/daolou/alfred-adcode/commit/ea791f278eba72d4a4a2343985ae6404e5c7457a))

### [1.2.2](https://github.com/daolou/alfred-adcode/compare/v1.2.1...v1.2.2) (2022-05-05)


### Chores | 其他更新(构建过程、辅助工具、库)

* keep Symbolic link ([b12d7cc](https://github.com/daolou/alfred-adcode/commit/b12d7ccdd42fea12f316663f9879317e1a4a3291))

### [1.2.1](https://github.com/daolou/alfred-adcode/compare/v1.2.0...v1.2.1) (2022-05-05)


### Chores | 其他更新(构建过程、辅助工具、库)

* build alfredworkflow ([174f612](https://github.com/daolou/alfred-adcode/commit/174f612be81a73a7b5183e070e639b4e7913b3ab))

## [1.2.0](https://github.com/daolou/alfred-adcode/compare/v1.1.0...v1.2.0) (2022-05-01)


### Features | 新功能

* ✨ exclude some file ([cc628e3](https://github.com/daolou/alfred-adcode/commit/cc628e3897ac4ebd471fdc1e606d077accdfeac3))

## [1.1.0](https://github.com/daolou/alfred-adcode/compare/v1.0.8...v1.1.0) (2022-04-30)


### Features | 新功能

* ✨ optimize Drastically reduced size ([587de2d](https://github.com/daolou/alfred-adcode/commit/587de2dd27699e44e3186359a448c26239f00b7e))

### [1.0.8](https://github.com/daolou/alfred-adcode/compare/v1.0.7...v1.0.8) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 pnpm i -P to optimize size ([65feec6](https://github.com/daolou/alfred-adcode/commit/65feec67db3fd89e70547d2f84c99bef65b5ad03))

### [1.0.7](https://github.com/daolou/alfred-adcode/compare/v1.0.6...v1.0.7) (2022-04-30)


### Bug Fixes | Bug 修复

* 🐛 No such file or directory ([061be58](https://github.com/daolou/alfred-adcode/commit/061be5805c059c7238f7f56042e8809164675b83))

### [1.0.6](https://github.com/daolou/alfred-adcode/compare/v1.0.5...v1.0.6) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 fix action build ([586eb8d](https://github.com/daolou/alfred-adcode/commit/586eb8deff7d7127372d2d2476ee1f197347258c))

### [1.0.5](https://github.com/daolou/alfred-adcode/compare/v1.0.4...v1.0.5) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 pnpm i --ignore-scripts ([8cda074](https://github.com/daolou/alfred-adcode/commit/8cda074f80bff53d9d8111f5709970b5c9edda86))

### [1.0.4](https://github.com/daolou/alfred-adcode/compare/v1.0.3...v1.0.4) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 pnpm env ([743186a](https://github.com/daolou/alfred-adcode/commit/743186a7c1f9a7fb1d2e42b29a7a716b8abc9d4e))

### [1.0.3](https://github.com/daolou/alfred-adcode/compare/v1.0.2...v1.0.3) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 ci cache use pnpm ([58744f7](https://github.com/daolou/alfred-adcode/commit/58744f76b047f178259323ee325d7c3797eec857))

### [1.0.2](https://github.com/daolou/alfred-adcode/compare/v1.0.1...v1.0.2) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 optimize build size ([ce96061](https://github.com/daolou/alfred-adcode/commit/ce960610f2092226f21e6bb8bf9e6d0aa2048597))

### [1.0.1](https://github.com/daolou/alfred-adcode/compare/v1.0.0...v1.0.1) (2022-04-30)


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 fix actions ([4bcf0a3](https://github.com/daolou/alfred-adcode/commit/4bcf0a3d348e7cf1f63d5e26ed4b550811c0f87b))

## 1.0.0 (2022-04-30)


### Bug Fixes | Bug 修复

* 🐛 ERR_MODULE_NOT_FOUND: node esm require extension .js and use cjs ([a8b82c2](https://github.com/daolou/alfred-adcode/commit/a8b82c23daa195a5004752dd97d6d38bad93b908))


### Documentation | 文档

* 📚️ Supplementary document ([0978d20](https://github.com/daolou/alfred-adcode/commit/0978d208616ef11138a08513e26ff3413a7b4fa1))


### Chores | 其他更新(构建过程、辅助工具、库)

* 🔧 add actions ([9e946e3](https://github.com/daolou/alfred-adcode/commit/9e946e3ea92af6a1f86b15dbca69ccd7fb405c9f))
* 🔧 first blood ([34587e4](https://github.com/daolou/alfred-adcode/commit/34587e490f89cb8550ef69ebeb7591d9dd176ba1))
* 🔧 pkg files include iconand info.plist ([e40e55f](https://github.com/daolou/alfred-adcode/commit/e40e55fe3af420986fa4e5aa83d81fcd44b69f40))
* 🔧 Remove the private key ([1547877](https://github.com/daolou/alfred-adcode/commit/154787768035fc5672666e47075fef00b8cca647))


src/exec.ts
/*
 * @Author: zhiguo.jzg
 * @Date: 2022-04-30 11:42:52
 * @Description: TODO: Description of file, its uses and information
 * @LastEditTime: 2022-04-30 11:55:27
 * @LastEditors: zhiguo.jzg
 */
import alfy from 'alfy';
import { getConfig, LIST_KEYS, LEVEL_MAP } from './config/index.js'; // node esm require extension .js

interface IDistrict {
  citycode: string;
  adcode: string;
  name: string;
  polyline?: string;
  center: string;
  level: string;
  districts?: IDistrict[];
}
interface IResponse {
  status: '0' | '1';
  info: string;
  infocode: string;
  count: number;
  districts: Array<IDistrict>;
}

export const run = async () => {
  try {
    const keywords = alfy.input;
    const { key, subdistrict, url } = getConfig();

    if (!key) {
      throw new Error('key is required');
    }
    const query = `?subdistrict=${subdistrict}&key=${key}&keywords=${keywords}`;
    const data = (await alfy.fetch(url + query)) as IResponse;
    if (data.status !== '1') {
      throw new Error(data.info);
    }
    const district = data.districts[0];
    if (!district) {
      throw new Error('no district found');
    }
    const items = LIST_KEYS.map(({ en, zh }) => {
      const value = String(district[en]||'');
      
      return {
        title: zh,
        subtitle: LEVEL_MAP[value] || value || '无',
        arg: LEVEL_MAP[value] || value,
        valid: true,
      };
    });
    alfy.output(items);
  } catch (err) {
    alfy.error(err as Error);
  }
};

index.js
/*
 * @Author: zhiguo.jzg
 * @Date: 2022-04-01 01:48:42
 * @Description: TODO: Description of file, its uses and information
 * @LastEditTime: 2022-05-19 22:13:42
 * @LastEditors: zhiguo.jzg
 */
import aw from './node_modules/@jsany/aw/lib/index.js';

const LIST_KEYS = [
  {
    en: 'name',
    zh: '行政区名称',
  },
  {
    en: 'adcode',
    zh: '区域编码',
  },
  {
    en: 'citycode',
    zh: '城市编码',
  },
  {
    en: 'center',
    zh: '区域中心点',
  },
  {
    en: 'level',
    zh: '行政区级别',
  },
  {
    en: 'polyline',
    zh: '行政区边界坐标点',
  },
];

const LEVEL_MAP = {
  country: '国家',

  province: '省份（直辖市会在province和city显示）',

  city: '市（直辖市会在province和city显示）',

  district: '区县',

  street: '街道',
};

const getConfig = () => ({
  url:
			aw.getEnv('ADCODE_API_URL')
			|| 'https://restapi.amap.com/v3/config/district',
  key: aw.getEnv('ADCODE_KEY'),
  subdistrict: aw.getEnv('ADCODE_SUBDISTRICT') || 0,
});

const adcode = async () => {
  try {
    const keywords = aw.input;
    const {key, subdistrict, url} = getConfig();

    if (!key) {
      throw new Error('key is required');
    }

    const query = `?subdistrict=${subdistrict}&key=${key}&keywords=${keywords}`;
    const {data} = await aw.fetch.get(url + query);
    if (data.status !== '1') {
      throw new Error(data.info);
    }

    const district = data.districts[0];
    if (!district) {
      throw new Error('no district found');
    }

    const items = LIST_KEYS.map(({en, zh}) => {
      const value = String(district[en] || '');

      return {
        title: zh,
        subtitle: LEVEL_MAP[value] || value || '无',
        arg: LEVEL_MAP[value] || value,
        valid: true,
      };
    });
    aw.output(items);
  } catch (error) {
    aw.error(error);
  }
};

adcode();
